markets/storageadapter/ondealsectorcommitted.go

```diff
@@ -0,0 +1,133 @@
+package storageadapter
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-fil-markets/storagemarket"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/lotus/build"
+	"github.com/filecoin-project/lotus/chain/actors/builtin/miner"
+	"github.com/filecoin-project/lotus/chain/events"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+type sectorCommittedEventsAPI interface {
+	Called(check events.CheckFunc, msgHnd events.MsgHandler, rev events.RevertHandler, confidence int, timeout abi.ChainEpoch, mf events.MsgMatchFunc) error
+}
+
+func OnDealSectorCommitted(ctx context.Context, api getCurrentDealInfoAPI, eventsApi sectorCommittedEventsAPI, provider address.Address, dealID abi.DealID, publishCid *cid.Cid, cb storagemarket.DealSectorCommittedCallback) error {
+	checkFunc := func(ts *types.TipSet) (done bool, more bool, err error) {
+		newDealID, sd, err := GetCurrentDealInfo(ctx, ts, api, dealID, publishCid)
+		if err != nil {
+			// TODO: This may be fine for some errors
+			return false, false, xerrors.Errorf("failed to look up deal on chain: %w", err)
+		}
+		dealID = newDealID
+
+		if sd.State.SectorStartEpoch > 0 {
+			cb(nil)
+			return true, false, nil
+		}
+
+		return false, true, nil
+	}
+
+	var sectorNumber abi.SectorNumber
+	var sectorFound bool
+
+	called := func(msg *types.Message, rec *types.MessageReceipt, ts *types.TipSet, curH abi.ChainEpoch) (more bool, err error) {
+		defer func() {
+			if err != nil {
+				cb(xerrors.Errorf("handling applied event: %w", err))
+			}
+		}()
+		switch msg.Method {
+		case miner.Methods.PreCommitSector:
+			dealID, _, err = GetCurrentDealInfo(ctx, ts, api, dealID, publishCid)
+			if err != nil {
+				return false, err
+			}
```

Comment:
Given that this function isn't exactly fast, we should:
* Place it after params.UnmarshalCBOR, and check len(params.DealIDs)>0
* Maybe cache last check based on ts.Key (bigger miners do send multiple precommits per epoch)